{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## 1. Load packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# data processing packages\n",
    "import numpy as np\n",
    "from numpy import random\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "# machine learning packages\n",
    "from sklearn.impute import SimpleImputer\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn import metrics\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "# visualization packages\n",
    "import seaborn as sb\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# other packages\n",
    "import torch\n",
    "from pathlib import Path\n",
    "import random\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## 2. Read the files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## Create the necessary folders\n",
    "Path('./Figures/').mkdir(parents=True, exist_ok=True)\n",
    "Path('./Results/').mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RefSeq_accession_number\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>AO-A12D.01TCGA</th>\n",
       "      <th>C8-A131.01TCGA</th>\n",
       "      <th>AO-A12B.01TCGA</th>\n",
       "      <th>BH-A18Q.02TCGA</th>\n",
       "      <th>C8-A130.02TCGA</th>\n",
       "      <th>C8-A138.03TCGA</th>\n",
       "      <th>E2-A154.03TCGA</th>\n",
       "      <th>C8-A12L.04TCGA</th>\n",
       "      <th>...</th>\n",
       "      <th>AO-A12B.34TCGA</th>\n",
       "      <th>A2-A0SW.35TCGA</th>\n",
       "      <th>AO-A0JL.35TCGA</th>\n",
       "      <th>BH-A0BV.35TCGA</th>\n",
       "      <th>A2-A0YM.36TCGA</th>\n",
       "      <th>BH-A0C7.36TCGA</th>\n",
       "      <th>A2-A0SX.36TCGA</th>\n",
       "      <th>263d3f-I.CPTAC</th>\n",
       "      <th>blcdb9-I.CPTAC</th>\n",
       "      <th>c4155b-C.CPTAC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RefSeq_accession_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_958782</th>\n",
       "      <td>PLEC</td>\n",
       "      <td>plectin isoform 1</td>\n",
       "      <td>1.096131</td>\n",
       "      <td>2.609943</td>\n",
       "      <td>-0.659828</td>\n",
       "      <td>0.195341</td>\n",
       "      <td>-0.494060</td>\n",
       "      <td>2.765081</td>\n",
       "      <td>0.862659</td>\n",
       "      <td>1.407570</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.963904</td>\n",
       "      <td>-0.487772</td>\n",
       "      <td>-0.10668</td>\n",
       "      <td>-0.065838</td>\n",
       "      <td>0.655850</td>\n",
       "      <td>-0.552212</td>\n",
       "      <td>-0.398560</td>\n",
       "      <td>0.598585</td>\n",
       "      <td>-0.191285</td>\n",
       "      <td>0.566975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_958785</th>\n",
       "      <td>NaN</td>\n",
       "      <td>plectin isoform 1g</td>\n",
       "      <td>1.111370</td>\n",
       "      <td>2.650422</td>\n",
       "      <td>-0.648742</td>\n",
       "      <td>0.215413</td>\n",
       "      <td>-0.503899</td>\n",
       "      <td>2.779709</td>\n",
       "      <td>0.870186</td>\n",
       "      <td>1.407570</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.938210</td>\n",
       "      <td>-0.487772</td>\n",
       "      <td>-0.10668</td>\n",
       "      <td>-0.055893</td>\n",
       "      <td>0.658143</td>\n",
       "      <td>-0.547749</td>\n",
       "      <td>-0.392601</td>\n",
       "      <td>0.606697</td>\n",
       "      <td>-0.183918</td>\n",
       "      <td>0.578702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_958786</th>\n",
       "      <td>PLEC</td>\n",
       "      <td>plectin isoform 1a</td>\n",
       "      <td>1.111370</td>\n",
       "      <td>2.650422</td>\n",
       "      <td>-0.654285</td>\n",
       "      <td>0.215413</td>\n",
       "      <td>-0.500619</td>\n",
       "      <td>2.779709</td>\n",
       "      <td>0.870186</td>\n",
       "      <td>1.410312</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.943919</td>\n",
       "      <td>-0.487772</td>\n",
       "      <td>-0.10668</td>\n",
       "      <td>-0.065838</td>\n",
       "      <td>0.655850</td>\n",
       "      <td>-0.552212</td>\n",
       "      <td>-0.392601</td>\n",
       "      <td>0.603993</td>\n",
       "      <td>-0.186022</td>\n",
       "      <td>0.576747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_000436</th>\n",
       "      <td>NaN</td>\n",
       "      <td>plectin isoform 1c</td>\n",
       "      <td>1.107561</td>\n",
       "      <td>2.646374</td>\n",
       "      <td>-0.632113</td>\n",
       "      <td>0.205377</td>\n",
       "      <td>-0.510459</td>\n",
       "      <td>2.797995</td>\n",
       "      <td>0.866423</td>\n",
       "      <td>1.407570</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.935355</td>\n",
       "      <td>-0.487772</td>\n",
       "      <td>-0.10668</td>\n",
       "      <td>-0.055893</td>\n",
       "      <td>0.655850</td>\n",
       "      <td>-0.552212</td>\n",
       "      <td>-0.392601</td>\n",
       "      <td>0.603993</td>\n",
       "      <td>-0.186022</td>\n",
       "      <td>0.576747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_958781</th>\n",
       "      <td>NaN</td>\n",
       "      <td>plectin isoform 1e</td>\n",
       "      <td>1.115180</td>\n",
       "      <td>2.646374</td>\n",
       "      <td>-0.640428</td>\n",
       "      <td>0.215413</td>\n",
       "      <td>-0.503899</td>\n",
       "      <td>2.787023</td>\n",
       "      <td>0.870186</td>\n",
       "      <td>1.413053</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.935355</td>\n",
       "      <td>-0.503853</td>\n",
       "      <td>-0.10668</td>\n",
       "      <td>-0.062523</td>\n",
       "      <td>0.651264</td>\n",
       "      <td>-0.556675</td>\n",
       "      <td>-0.395581</td>\n",
       "      <td>0.603993</td>\n",
       "      <td>-0.167079</td>\n",
       "      <td>0.576747</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 85 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gene_symbol           gene_name  AO-A12D.01TCGA  \\\n",
       "RefSeq_accession_number                                                   \n",
       "NP_958782                      PLEC   plectin isoform 1        1.096131   \n",
       "NP_958785                       NaN  plectin isoform 1g        1.111370   \n",
       "NP_958786                      PLEC  plectin isoform 1a        1.111370   \n",
       "NP_000436                       NaN  plectin isoform 1c        1.107561   \n",
       "NP_958781                       NaN  plectin isoform 1e        1.115180   \n",
       "\n",
       "                         C8-A131.01TCGA  AO-A12B.01TCGA  BH-A18Q.02TCGA  \\\n",
       "RefSeq_accession_number                                                   \n",
       "NP_958782                      2.609943       -0.659828        0.195341   \n",
       "NP_958785                      2.650422       -0.648742        0.215413   \n",
       "NP_958786                      2.650422       -0.654285        0.215413   \n",
       "NP_000436                      2.646374       -0.632113        0.205377   \n",
       "NP_958781                      2.646374       -0.640428        0.215413   \n",
       "\n",
       "                         C8-A130.02TCGA  C8-A138.03TCGA  E2-A154.03TCGA  \\\n",
       "RefSeq_accession_number                                                   \n",
       "NP_958782                     -0.494060        2.765081        0.862659   \n",
       "NP_958785                     -0.503899        2.779709        0.870186   \n",
       "NP_958786                     -0.500619        2.779709        0.870186   \n",
       "NP_000436                     -0.510459        2.797995        0.866423   \n",
       "NP_958781                     -0.503899        2.787023        0.870186   \n",
       "\n",
       "                         C8-A12L.04TCGA  ...  AO-A12B.34TCGA  A2-A0SW.35TCGA  \\\n",
       "RefSeq_accession_number                  ...                                   \n",
       "NP_958782                      1.407570  ...       -0.963904       -0.487772   \n",
       "NP_958785                      1.407570  ...       -0.938210       -0.487772   \n",
       "NP_958786                      1.410312  ...       -0.943919       -0.487772   \n",
       "NP_000436                      1.407570  ...       -0.935355       -0.487772   \n",
       "NP_958781                      1.413053  ...       -0.935355       -0.503853   \n",
       "\n",
       "                         AO-A0JL.35TCGA  BH-A0BV.35TCGA  A2-A0YM.36TCGA  \\\n",
       "RefSeq_accession_number                                                   \n",
       "NP_958782                      -0.10668       -0.065838        0.655850   \n",
       "NP_958785                      -0.10668       -0.055893        0.658143   \n",
       "NP_958786                      -0.10668       -0.065838        0.655850   \n",
       "NP_000436                      -0.10668       -0.055893        0.655850   \n",
       "NP_958781                      -0.10668       -0.062523        0.651264   \n",
       "\n",
       "                         BH-A0C7.36TCGA  A2-A0SX.36TCGA  263d3f-I.CPTAC  \\\n",
       "RefSeq_accession_number                                                   \n",
       "NP_958782                     -0.552212       -0.398560        0.598585   \n",
       "NP_958785                     -0.547749       -0.392601        0.606697   \n",
       "NP_958786                     -0.552212       -0.392601        0.603993   \n",
       "NP_000436                     -0.552212       -0.392601        0.603993   \n",
       "NP_958781                     -0.556675       -0.395581        0.603993   \n",
       "\n",
       "                         blcdb9-I.CPTAC  c4155b-C.CPTAC  \n",
       "RefSeq_accession_number                                  \n",
       "NP_958782                     -0.191285        0.566975  \n",
       "NP_958785                     -0.183918        0.578702  \n",
       "NP_958786                     -0.186022        0.576747  \n",
       "NP_000436                     -0.186022        0.576747  \n",
       "NP_958781                     -0.167079        0.576747  \n",
       "\n",
       "[5 rows x 85 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Set path to the data set\n",
    "dataset_path = \"./dataset/77_cancer_proteomes_CPTAC_itraq.csv\"\n",
    "clinical_info = \"./dataset/clinical_data_breast_cancer.csv\"\n",
    "pam50_proteins = \"./dataset/PAM50_proteins.csv\"\n",
    "\n",
    "## Load data\n",
    "data = pd.read_csv(dataset_path,header=0,index_col=0)\n",
    "clinical_file = pd.read_csv(clinical_info,header=0,index_col=0) ## holds clinical information about each patient/sample\n",
    "pam50 = pd.read_csv(pam50_proteins,header=0)\n",
    "\n",
    "# RefSeq protein ID (each protein has a unique ID in a RefSeq database)\n",
    "print(data.index.name)\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## 3. Data Set Processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-AO-A12D</th>\n",
       "      <th>TCGA-C8-A131</th>\n",
       "      <th>TCGA-AO-A12B</th>\n",
       "      <th>TCGA-BH-A18Q</th>\n",
       "      <th>TCGA-C8-A130</th>\n",
       "      <th>TCGA-C8-A138</th>\n",
       "      <th>TCGA-E2-A154</th>\n",
       "      <th>TCGA-C8-A12L</th>\n",
       "      <th>TCGA-A2-A0EX</th>\n",
       "      <th>TCGA-AO-A12D</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-AO-A12B</th>\n",
       "      <th>TCGA-A2-A0SW</th>\n",
       "      <th>TCGA-AO-A0JL</th>\n",
       "      <th>TCGA-BH-A0BV</th>\n",
       "      <th>TCGA-A2-A0YM</th>\n",
       "      <th>TCGA-BH-A0C7</th>\n",
       "      <th>TCGA-A2-A0SX</th>\n",
       "      <th>263d3f-I.CPTAC</th>\n",
       "      <th>blcdb9-I.CPTAC</th>\n",
       "      <th>c4155b-C.CPTAC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RefSeq_accession_number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NP_958782</th>\n",
       "      <td>1.096131</td>\n",
       "      <td>2.609943</td>\n",
       "      <td>-0.659828</td>\n",
       "      <td>0.195341</td>\n",
       "      <td>-0.494060</td>\n",
       "      <td>2.765081</td>\n",
       "      <td>0.862659</td>\n",
       "      <td>1.407570</td>\n",
       "      <td>1.185108</td>\n",
       "      <td>1.100688</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.963904</td>\n",
       "      <td>-0.487772</td>\n",
       "      <td>-0.10668</td>\n",
       "      <td>-0.065838</td>\n",
       "      <td>0.655850</td>\n",
       "      <td>-0.552212</td>\n",
       "      <td>-0.398560</td>\n",
       "      <td>0.598585</td>\n",
       "      <td>-0.191285</td>\n",
       "      <td>0.566975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_958785</th>\n",
       "      <td>1.111370</td>\n",
       "      <td>2.650422</td>\n",
       "      <td>-0.648742</td>\n",
       "      <td>0.215413</td>\n",
       "      <td>-0.503899</td>\n",
       "      <td>2.779709</td>\n",
       "      <td>0.870186</td>\n",
       "      <td>1.407570</td>\n",
       "      <td>1.192612</td>\n",
       "      <td>1.100688</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.938210</td>\n",
       "      <td>-0.487772</td>\n",
       "      <td>-0.10668</td>\n",
       "      <td>-0.055893</td>\n",
       "      <td>0.658143</td>\n",
       "      <td>-0.547749</td>\n",
       "      <td>-0.392601</td>\n",
       "      <td>0.606697</td>\n",
       "      <td>-0.183918</td>\n",
       "      <td>0.578702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_958786</th>\n",
       "      <td>1.111370</td>\n",
       "      <td>2.650422</td>\n",
       "      <td>-0.654285</td>\n",
       "      <td>0.215413</td>\n",
       "      <td>-0.500619</td>\n",
       "      <td>2.779709</td>\n",
       "      <td>0.870186</td>\n",
       "      <td>1.410312</td>\n",
       "      <td>1.188860</td>\n",
       "      <td>1.100688</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.943919</td>\n",
       "      <td>-0.487772</td>\n",
       "      <td>-0.10668</td>\n",
       "      <td>-0.065838</td>\n",
       "      <td>0.655850</td>\n",
       "      <td>-0.552212</td>\n",
       "      <td>-0.392601</td>\n",
       "      <td>0.603993</td>\n",
       "      <td>-0.186022</td>\n",
       "      <td>0.576747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_000436</th>\n",
       "      <td>1.107561</td>\n",
       "      <td>2.646374</td>\n",
       "      <td>-0.632113</td>\n",
       "      <td>0.205377</td>\n",
       "      <td>-0.510459</td>\n",
       "      <td>2.797995</td>\n",
       "      <td>0.866423</td>\n",
       "      <td>1.407570</td>\n",
       "      <td>1.185108</td>\n",
       "      <td>1.100688</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.935355</td>\n",
       "      <td>-0.487772</td>\n",
       "      <td>-0.10668</td>\n",
       "      <td>-0.055893</td>\n",
       "      <td>0.655850</td>\n",
       "      <td>-0.552212</td>\n",
       "      <td>-0.392601</td>\n",
       "      <td>0.603993</td>\n",
       "      <td>-0.186022</td>\n",
       "      <td>0.576747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NP_958781</th>\n",
       "      <td>1.115180</td>\n",
       "      <td>2.646374</td>\n",
       "      <td>-0.640428</td>\n",
       "      <td>0.215413</td>\n",
       "      <td>-0.503899</td>\n",
       "      <td>2.787023</td>\n",
       "      <td>0.870186</td>\n",
       "      <td>1.413053</td>\n",
       "      <td>1.200116</td>\n",
       "      <td>1.093358</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.935355</td>\n",
       "      <td>-0.503853</td>\n",
       "      <td>-0.10668</td>\n",
       "      <td>-0.062523</td>\n",
       "      <td>0.651264</td>\n",
       "      <td>-0.556675</td>\n",
       "      <td>-0.395581</td>\n",
       "      <td>0.603993</td>\n",
       "      <td>-0.167079</td>\n",
       "      <td>0.576747</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 83 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         TCGA-AO-A12D  TCGA-C8-A131  TCGA-AO-A12B  \\\n",
       "RefSeq_accession_number                                             \n",
       "NP_958782                    1.096131      2.609943     -0.659828   \n",
       "NP_958785                    1.111370      2.650422     -0.648742   \n",
       "NP_958786                    1.111370      2.650422     -0.654285   \n",
       "NP_000436                    1.107561      2.646374     -0.632113   \n",
       "NP_958781                    1.115180      2.646374     -0.640428   \n",
       "\n",
       "                         TCGA-BH-A18Q  TCGA-C8-A130  TCGA-C8-A138  \\\n",
       "RefSeq_accession_number                                             \n",
       "NP_958782                    0.195341     -0.494060      2.765081   \n",
       "NP_958785                    0.215413     -0.503899      2.779709   \n",
       "NP_958786                    0.215413     -0.500619      2.779709   \n",
       "NP_000436                    0.205377     -0.510459      2.797995   \n",
       "NP_958781                    0.215413     -0.503899      2.787023   \n",
       "\n",
       "                         TCGA-E2-A154  TCGA-C8-A12L  TCGA-A2-A0EX  \\\n",
       "RefSeq_accession_number                                             \n",
       "NP_958782                    0.862659      1.407570      1.185108   \n",
       "NP_958785                    0.870186      1.407570      1.192612   \n",
       "NP_958786                    0.870186      1.410312      1.188860   \n",
       "NP_000436                    0.866423      1.407570      1.185108   \n",
       "NP_958781                    0.870186      1.413053      1.200116   \n",
       "\n",
       "                         TCGA-AO-A12D  ...  TCGA-AO-A12B  TCGA-A2-A0SW  \\\n",
       "RefSeq_accession_number                ...                               \n",
       "NP_958782                    1.100688  ...     -0.963904     -0.487772   \n",
       "NP_958785                    1.100688  ...     -0.938210     -0.487772   \n",
       "NP_958786                    1.100688  ...     -0.943919     -0.487772   \n",
       "NP_000436                    1.100688  ...     -0.935355     -0.487772   \n",
       "NP_958781                    1.093358  ...     -0.935355     -0.503853   \n",
       "\n",
       "                         TCGA-AO-A0JL  TCGA-BH-A0BV  TCGA-A2-A0YM  \\\n",
       "RefSeq_accession_number                                             \n",
       "NP_958782                    -0.10668     -0.065838      0.655850   \n",
       "NP_958785                    -0.10668     -0.055893      0.658143   \n",
       "NP_958786                    -0.10668     -0.065838      0.655850   \n",
       "NP_000436                    -0.10668     -0.055893      0.655850   \n",
       "NP_958781                    -0.10668     -0.062523      0.651264   \n",
       "\n",
       "                         TCGA-BH-A0C7  TCGA-A2-A0SX  263d3f-I.CPTAC  \\\n",
       "RefSeq_accession_number                                               \n",
       "NP_958782                   -0.552212     -0.398560        0.598585   \n",
       "NP_958785                   -0.547749     -0.392601        0.606697   \n",
       "NP_958786                   -0.552212     -0.392601        0.603993   \n",
       "NP_000436                   -0.552212     -0.392601        0.603993   \n",
       "NP_958781                   -0.556675     -0.395581        0.603993   \n",
       "\n",
       "                         blcdb9-I.CPTAC  c4155b-C.CPTAC  \n",
       "RefSeq_accession_number                                  \n",
       "NP_958782                     -0.191285        0.566975  \n",
       "NP_958785                     -0.183918        0.578702  \n",
       "NP_958786                     -0.186022        0.576747  \n",
       "NP_000436                     -0.186022        0.576747  \n",
       "NP_958781                     -0.167079        0.576747  \n",
       "\n",
       "[5 rows x 83 columns]"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Drop unused information columns\n",
    "data.drop(['gene_symbol','gene_name'],axis=1,inplace=True)\n",
    "\n",
    "## Change the protein data sample names to a format matching the clinical data set\n",
    "data.rename(columns=lambda x: \"TCGA-%s\" % (re.split('[_|-|.]',x)[0]) if bool(re.search(\"TCGA\",x)) is True else x,inplace=True)\n",
    "\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12553, 83)\n",
      "(83, 12553)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>RefSeq_accession_number</th>\n",
       "      <th>NP_958782</th>\n",
       "      <th>NP_958785</th>\n",
       "      <th>NP_958786</th>\n",
       "      <th>NP_000436</th>\n",
       "      <th>NP_958781</th>\n",
       "      <th>NP_958780</th>\n",
       "      <th>NP_958783</th>\n",
       "      <th>NP_958784</th>\n",
       "      <th>NP_112598</th>\n",
       "      <th>NP_001611</th>\n",
       "      <th>...</th>\n",
       "      <th>NP_001193600</th>\n",
       "      <th>NP_061134</th>\n",
       "      <th>NP_932347</th>\n",
       "      <th>NP_003593</th>\n",
       "      <th>NP_997203</th>\n",
       "      <th>NP_001191293</th>\n",
       "      <th>NP_775791</th>\n",
       "      <th>NP_004065</th>\n",
       "      <th>NP_068752</th>\n",
       "      <th>NP_219494</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-AO-A12D</th>\n",
       "      <td>1.096131</td>\n",
       "      <td>1.111370</td>\n",
       "      <td>1.111370</td>\n",
       "      <td>1.107561</td>\n",
       "      <td>1.115180</td>\n",
       "      <td>1.107561</td>\n",
       "      <td>1.111370</td>\n",
       "      <td>1.111370</td>\n",
       "      <td>-1.517390</td>\n",
       "      <td>0.482754</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.340163</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.633517</td>\n",
       "      <td>12.666488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-C8-A131</th>\n",
       "      <td>2.609943</td>\n",
       "      <td>2.650422</td>\n",
       "      <td>2.650422</td>\n",
       "      <td>2.646374</td>\n",
       "      <td>2.646374</td>\n",
       "      <td>2.646374</td>\n",
       "      <td>2.650422</td>\n",
       "      <td>2.650422</td>\n",
       "      <td>3.909313</td>\n",
       "      <td>-1.045294</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.451902</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.840325</td>\n",
       "      <td>0.140736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-AO-A12B</th>\n",
       "      <td>-0.659828</td>\n",
       "      <td>-0.648742</td>\n",
       "      <td>-0.654285</td>\n",
       "      <td>-0.632113</td>\n",
       "      <td>-0.640428</td>\n",
       "      <td>-0.654285</td>\n",
       "      <td>-0.648742</td>\n",
       "      <td>-0.648742</td>\n",
       "      <td>-0.618256</td>\n",
       "      <td>1.222003</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.718531</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.965192</td>\n",
       "      <td>-2.854835</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-BH-A18Q</th>\n",
       "      <td>0.195341</td>\n",
       "      <td>0.215413</td>\n",
       "      <td>0.215413</td>\n",
       "      <td>0.205377</td>\n",
       "      <td>0.215413</td>\n",
       "      <td>0.215413</td>\n",
       "      <td>0.215413</td>\n",
       "      <td>0.215413</td>\n",
       "      <td>-1.035760</td>\n",
       "      <td>-0.517226</td>\n",
       "      <td>...</td>\n",
       "      <td>0.048144</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.881872</td>\n",
       "      <td>2.527072</td>\n",
       "      <td>-8.111243</td>\n",
       "      <td>-16.029761</td>\n",
       "      <td>-2.046065</td>\n",
       "      <td>-1.778435</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-3.069752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-C8-A130</th>\n",
       "      <td>-0.494060</td>\n",
       "      <td>-0.503899</td>\n",
       "      <td>-0.500619</td>\n",
       "      <td>-0.510459</td>\n",
       "      <td>-0.503899</td>\n",
       "      <td>-0.503899</td>\n",
       "      <td>-0.500619</td>\n",
       "      <td>-0.500619</td>\n",
       "      <td>-1.845366</td>\n",
       "      <td>-0.405503</td>\n",
       "      <td>...</td>\n",
       "      <td>1.457462</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.710012</td>\n",
       "      <td>0.296389</td>\n",
       "      <td>-1.753529</td>\n",
       "      <td>1.729692</td>\n",
       "      <td>-0.425182</td>\n",
       "      <td>-0.149673</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.047997</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 12553 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "RefSeq_accession_number  NP_958782  NP_958785  NP_958786  NP_000436  \\\n",
       "TCGA-AO-A12D              1.096131   1.111370   1.111370   1.107561   \n",
       "TCGA-C8-A131              2.609943   2.650422   2.650422   2.646374   \n",
       "TCGA-AO-A12B             -0.659828  -0.648742  -0.654285  -0.632113   \n",
       "TCGA-BH-A18Q              0.195341   0.215413   0.215413   0.205377   \n",
       "TCGA-C8-A130             -0.494060  -0.503899  -0.500619  -0.510459   \n",
       "\n",
       "RefSeq_accession_number  NP_958781  NP_958780  NP_958783  NP_958784  \\\n",
       "TCGA-AO-A12D              1.115180   1.107561   1.111370   1.111370   \n",
       "TCGA-C8-A131              2.646374   2.646374   2.650422   2.650422   \n",
       "TCGA-AO-A12B             -0.640428  -0.654285  -0.648742  -0.648742   \n",
       "TCGA-BH-A18Q              0.215413   0.215413   0.215413   0.215413   \n",
       "TCGA-C8-A130             -0.503899  -0.503899  -0.500619  -0.500619   \n",
       "\n",
       "RefSeq_accession_number  NP_112598  NP_001611  ...  NP_001193600  NP_061134  \\\n",
       "TCGA-AO-A12D             -1.517390   0.482754  ...           NaN        NaN   \n",
       "TCGA-C8-A131              3.909313  -1.045294  ...           NaN        NaN   \n",
       "TCGA-AO-A12B             -0.618256   1.222003  ...           NaN        NaN   \n",
       "TCGA-BH-A18Q             -1.035760  -0.517226  ...      0.048144        NaN   \n",
       "TCGA-C8-A130             -1.845366  -0.405503  ...      1.457462        NaN   \n",
       "\n",
       "RefSeq_accession_number  NP_932347  NP_003593  NP_997203  NP_001191293  \\\n",
       "TCGA-AO-A12D                   NaN  -0.340163        NaN           NaN   \n",
       "TCGA-C8-A131                   NaN   3.451902        NaN           NaN   \n",
       "TCGA-AO-A12B                   NaN  -1.718531        NaN           NaN   \n",
       "TCGA-BH-A18Q             -0.881872   2.527072  -8.111243    -16.029761   \n",
       "TCGA-C8-A130              1.710012   0.296389  -1.753529      1.729692   \n",
       "\n",
       "RefSeq_accession_number  NP_775791  NP_004065  NP_068752  NP_219494  \n",
       "TCGA-AO-A12D                   NaN        NaN  -0.633517  12.666488  \n",
       "TCGA-C8-A131                   NaN        NaN   4.840325   0.140736  \n",
       "TCGA-AO-A12B                   NaN        NaN  -1.965192  -2.854835  \n",
       "TCGA-BH-A18Q             -2.046065  -1.778435        NaN  -3.069752  \n",
       "TCGA-C8-A130             -0.425182  -0.149673        NaN  -0.047997  \n",
       "\n",
       "[5 rows x 12553 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Transpose data for the clustering algorithm since we want to divide patient samples, not proteins\n",
    "print(data.shape)\n",
    "datat = data.transpose()\n",
    "print(datat.shape)\n",
    "\n",
    "datat.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of patients in clinical data set:  105\n",
      "Number of patients in protein data set:  83\n"
     ]
    }
   ],
   "source": [
    "print(\"Number of patients in clinical data set: \", len(clinical_file.index))\n",
    "print(\"Number of patients in protein data set: \", len(datat.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The shape of the clinical data set:  (77, 29)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gender</th>\n",
       "      <th>Age at Initial Pathologic Diagnosis</th>\n",
       "      <th>ER Status</th>\n",
       "      <th>PR Status</th>\n",
       "      <th>HER2 Final Status</th>\n",
       "      <th>Tumor</th>\n",
       "      <th>Tumor--T1 Coded</th>\n",
       "      <th>Node</th>\n",
       "      <th>Node-Coded</th>\n",
       "      <th>Metastasis</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM50 mRNA</th>\n",
       "      <th>SigClust Unsupervised mRNA</th>\n",
       "      <th>SigClust Intrinsic mRNA</th>\n",
       "      <th>miRNA Clusters</th>\n",
       "      <th>methylation Clusters</th>\n",
       "      <th>RPPA Clusters</th>\n",
       "      <th>CN Clusters</th>\n",
       "      <th>Integrated Clusters (with PAM50)</th>\n",
       "      <th>Integrated Clusters (no exp)</th>\n",
       "      <th>Integrated Clusters (unsup exp)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Complete TCGA ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0CM</th>\n",
       "      <td>FEMALE</td>\n",
       "      <td>40</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>T2</td>\n",
       "      <td>T_Other</td>\n",
       "      <td>N0</td>\n",
       "      <td>Negative</td>\n",
       "      <td>M0</td>\n",
       "      <td>...</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>-12</td>\n",
       "      <td>-13</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>Basal</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-BH-A18Q</th>\n",
       "      <td>FEMALE</td>\n",
       "      <td>56</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>T2</td>\n",
       "      <td>T_Other</td>\n",
       "      <td>N1</td>\n",
       "      <td>Positive</td>\n",
       "      <td>M0</td>\n",
       "      <td>...</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>-12</td>\n",
       "      <td>-13</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>Basal</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A7-A0CE</th>\n",
       "      <td>FEMALE</td>\n",
       "      <td>57</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>T2</td>\n",
       "      <td>T_Other</td>\n",
       "      <td>N0</td>\n",
       "      <td>Negative</td>\n",
       "      <td>M0</td>\n",
       "      <td>...</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>0</td>\n",
       "      <td>-13</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>Basal</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-D8-A142</th>\n",
       "      <td>FEMALE</td>\n",
       "      <td>74</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>T3</td>\n",
       "      <td>T_Other</td>\n",
       "      <td>N0</td>\n",
       "      <td>Negative</td>\n",
       "      <td>M0</td>\n",
       "      <td>...</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>0</td>\n",
       "      <td>-13</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>X</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-AO-A0J6</th>\n",
       "      <td>FEMALE</td>\n",
       "      <td>61</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>Negative</td>\n",
       "      <td>T2</td>\n",
       "      <td>T_Other</td>\n",
       "      <td>N0</td>\n",
       "      <td>Negative</td>\n",
       "      <td>M0</td>\n",
       "      <td>...</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>-12</td>\n",
       "      <td>-13</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>Basal</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Gender  Age at Initial Pathologic Diagnosis ER Status  \\\n",
       "Complete TCGA ID                                                          \n",
       "TCGA-A2-A0CM      FEMALE                                   40  Negative   \n",
       "TCGA-BH-A18Q      FEMALE                                   56  Negative   \n",
       "TCGA-A7-A0CE      FEMALE                                   57  Negative   \n",
       "TCGA-D8-A142      FEMALE                                   74  Negative   \n",
       "TCGA-AO-A0J6      FEMALE                                   61  Negative   \n",
       "\n",
       "                 PR Status HER2 Final Status Tumor Tumor--T1 Coded Node  \\\n",
       "Complete TCGA ID                                                          \n",
       "TCGA-A2-A0CM      Negative          Negative    T2         T_Other   N0   \n",
       "TCGA-BH-A18Q      Negative          Negative    T2         T_Other   N1   \n",
       "TCGA-A7-A0CE      Negative          Negative    T2         T_Other   N0   \n",
       "TCGA-D8-A142      Negative          Negative    T3         T_Other   N0   \n",
       "TCGA-AO-A0J6      Negative          Negative    T2         T_Other   N0   \n",
       "\n",
       "                 Node-Coded Metastasis  ...  PAM50 mRNA  \\\n",
       "Complete TCGA ID                        ...               \n",
       "TCGA-A2-A0CM       Negative         M0  ...  Basal-like   \n",
       "TCGA-BH-A18Q       Positive         M0  ...  Basal-like   \n",
       "TCGA-A7-A0CE       Negative         M0  ...  Basal-like   \n",
       "TCGA-D8-A142       Negative         M0  ...  Basal-like   \n",
       "TCGA-AO-A0J6       Negative         M0  ...  Basal-like   \n",
       "\n",
       "                 SigClust Unsupervised mRNA SigClust Intrinsic mRNA  \\\n",
       "Complete TCGA ID                                                      \n",
       "TCGA-A2-A0CM                            -12                     -13   \n",
       "TCGA-BH-A18Q                            -12                     -13   \n",
       "TCGA-A7-A0CE                              0                     -13   \n",
       "TCGA-D8-A142                              0                     -13   \n",
       "TCGA-AO-A0J6                            -12                     -13   \n",
       "\n",
       "                 miRNA Clusters methylation Clusters  RPPA Clusters  \\\n",
       "Complete TCGA ID                                                      \n",
       "TCGA-A2-A0CM                  4                    4          Basal   \n",
       "TCGA-BH-A18Q                  5                    5          Basal   \n",
       "TCGA-A7-A0CE                  5                    5          Basal   \n",
       "TCGA-D8-A142                  3                    5              X   \n",
       "TCGA-AO-A0J6                  2                    5          Basal   \n",
       "\n",
       "                  CN Clusters  Integrated Clusters (with PAM50)  \\\n",
       "Complete TCGA ID                                                  \n",
       "TCGA-A2-A0CM                4                                 2   \n",
       "TCGA-BH-A18Q                1                                 2   \n",
       "TCGA-A7-A0CE                1                                 2   \n",
       "TCGA-D8-A142                1                                 2   \n",
       "TCGA-AO-A0J6                1                                 2   \n",
       "\n",
       "                  Integrated Clusters (no exp) Integrated Clusters (unsup exp)  \n",
       "Complete TCGA ID                                                                \n",
       "TCGA-A2-A0CM                                 1                               1  \n",
       "TCGA-BH-A18Q                                 2                               2  \n",
       "TCGA-A7-A0CE                                 2                               2  \n",
       "TCGA-D8-A142                                 2                               2  \n",
       "TCGA-AO-A0J6                                 2                               2  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Drop clinical entries for samples not in our protein data set\n",
    "clinical = clinical_file.loc[[x for x in clinical_file.index.tolist() if x in datat.index],:]\n",
    "\n",
    "print(\"The shape of the clinical data set: \", clinical.shape)\n",
    "clinical.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape of the merged data set:  (74, 12582)\n",
      "with 74 patients and 12582 features\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NP_958782</th>\n",
       "      <th>NP_958785</th>\n",
       "      <th>NP_958786</th>\n",
       "      <th>NP_000436</th>\n",
       "      <th>NP_958781</th>\n",
       "      <th>NP_958780</th>\n",
       "      <th>NP_958783</th>\n",
       "      <th>NP_958784</th>\n",
       "      <th>NP_112598</th>\n",
       "      <th>NP_001611</th>\n",
       "      <th>...</th>\n",
       "      <th>PAM50 mRNA</th>\n",
       "      <th>SigClust Unsupervised mRNA</th>\n",
       "      <th>SigClust Intrinsic mRNA</th>\n",
       "      <th>miRNA Clusters</th>\n",
       "      <th>methylation Clusters</th>\n",
       "      <th>RPPA Clusters</th>\n",
       "      <th>CN Clusters</th>\n",
       "      <th>Integrated Clusters (with PAM50)</th>\n",
       "      <th>Integrated Clusters (no exp)</th>\n",
       "      <th>Integrated Clusters (unsup exp)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0CM</th>\n",
       "      <td>0.683404</td>\n",
       "      <td>0.694424</td>\n",
       "      <td>0.698098</td>\n",
       "      <td>0.687077</td>\n",
       "      <td>0.687077</td>\n",
       "      <td>0.698098</td>\n",
       "      <td>0.698098</td>\n",
       "      <td>0.698098</td>\n",
       "      <td>-2.652150</td>\n",
       "      <td>-0.984373</td>\n",
       "      <td>...</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>-12</td>\n",
       "      <td>-13</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>Basal</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0D2</th>\n",
       "      <td>0.107491</td>\n",
       "      <td>0.104164</td>\n",
       "      <td>0.107491</td>\n",
       "      <td>0.097512</td>\n",
       "      <td>0.104164</td>\n",
       "      <td>0.104164</td>\n",
       "      <td>0.104164</td>\n",
       "      <td>0.104164</td>\n",
       "      <td>-0.880454</td>\n",
       "      <td>-1.512473</td>\n",
       "      <td>...</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>-12</td>\n",
       "      <td>-13</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>Basal</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0EQ</th>\n",
       "      <td>-0.912670</td>\n",
       "      <td>-0.927979</td>\n",
       "      <td>-0.927979</td>\n",
       "      <td>-0.931806</td>\n",
       "      <td>-0.927979</td>\n",
       "      <td>-0.927979</td>\n",
       "      <td>-0.927979</td>\n",
       "      <td>-0.927979</td>\n",
       "      <td>-3.071151</td>\n",
       "      <td>-2.278943</td>\n",
       "      <td>...</td>\n",
       "      <td>HER2-enriched</td>\n",
       "      <td>-5</td>\n",
       "      <td>-2</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>Basal</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0EV</th>\n",
       "      <td>0.452986</td>\n",
       "      <td>0.472590</td>\n",
       "      <td>0.472590</td>\n",
       "      <td>0.458587</td>\n",
       "      <td>0.472590</td>\n",
       "      <td>0.472590</td>\n",
       "      <td>0.472590</td>\n",
       "      <td>0.472590</td>\n",
       "      <td>-0.742871</td>\n",
       "      <td>1.811277</td>\n",
       "      <td>...</td>\n",
       "      <td>Luminal A</td>\n",
       "      <td>-4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>ReacI</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0EX</th>\n",
       "      <td>1.185108</td>\n",
       "      <td>1.192612</td>\n",
       "      <td>1.188860</td>\n",
       "      <td>1.185108</td>\n",
       "      <td>1.200116</td>\n",
       "      <td>1.188860</td>\n",
       "      <td>1.188860</td>\n",
       "      <td>1.192612</td>\n",
       "      <td>1.046289</td>\n",
       "      <td>2.138081</td>\n",
       "      <td>...</td>\n",
       "      <td>Luminal A</td>\n",
       "      <td>-7</td>\n",
       "      <td>-5</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>ReacI</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 12582 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              NP_958782  NP_958785  NP_958786  NP_000436  NP_958781  \\\n",
       "TCGA-A2-A0CM   0.683404   0.694424   0.698098   0.687077   0.687077   \n",
       "TCGA-A2-A0D2   0.107491   0.104164   0.107491   0.097512   0.104164   \n",
       "TCGA-A2-A0EQ  -0.912670  -0.927979  -0.927979  -0.931806  -0.927979   \n",
       "TCGA-A2-A0EV   0.452986   0.472590   0.472590   0.458587   0.472590   \n",
       "TCGA-A2-A0EX   1.185108   1.192612   1.188860   1.185108   1.200116   \n",
       "\n",
       "              NP_958780  NP_958783  NP_958784  NP_112598  NP_001611  ...  \\\n",
       "TCGA-A2-A0CM   0.698098   0.698098   0.698098  -2.652150  -0.984373  ...   \n",
       "TCGA-A2-A0D2   0.104164   0.104164   0.104164  -0.880454  -1.512473  ...   \n",
       "TCGA-A2-A0EQ  -0.927979  -0.927979  -0.927979  -3.071151  -2.278943  ...   \n",
       "TCGA-A2-A0EV   0.472590   0.472590   0.472590  -0.742871   1.811277  ...   \n",
       "TCGA-A2-A0EX   1.188860   1.188860   1.192612   1.046289   2.138081  ...   \n",
       "\n",
       "                 PAM50 mRNA  SigClust Unsupervised mRNA  \\\n",
       "TCGA-A2-A0CM     Basal-like                         -12   \n",
       "TCGA-A2-A0D2     Basal-like                         -12   \n",
       "TCGA-A2-A0EQ  HER2-enriched                          -5   \n",
       "TCGA-A2-A0EV      Luminal A                          -4   \n",
       "TCGA-A2-A0EX      Luminal A                          -7   \n",
       "\n",
       "              SigClust Intrinsic mRNA  miRNA Clusters  methylation Clusters  \\\n",
       "TCGA-A2-A0CM                      -13               4                     4   \n",
       "TCGA-A2-A0D2                      -13               4                     5   \n",
       "TCGA-A2-A0EQ                       -2               5                     4   \n",
       "TCGA-A2-A0EV                        0               4                     2   \n",
       "TCGA-A2-A0EX                       -5               4                     4   \n",
       "\n",
       "              RPPA Clusters  CN Clusters  Integrated Clusters (with PAM50)  \\\n",
       "TCGA-A2-A0CM          Basal            4                                 2   \n",
       "TCGA-A2-A0D2          Basal            3                                 2   \n",
       "TCGA-A2-A0EQ          Basal            4                                 4   \n",
       "TCGA-A2-A0EV          ReacI            3                                 3   \n",
       "TCGA-A2-A0EX          ReacI            4                                 3   \n",
       "\n",
       "              Integrated Clusters (no exp)  Integrated Clusters (unsup exp)  \n",
       "TCGA-A2-A0CM                             1                                1  \n",
       "TCGA-A2-A0D2                             2                                2  \n",
       "TCGA-A2-A0EQ                             1                                1  \n",
       "TCGA-A2-A0EV                             3                                4  \n",
       "TCGA-A2-A0EX                             1                                4  \n",
       "\n",
       "[5 rows x 12582 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Add clinical meta data to our protein data set, note: all numerical features for analysis start with NP_ or XP_\n",
    "merged = datat.merge(clinical,left_index=True,right_index=True)\n",
    "\n",
    "# Drop the duplicated columns (added by Pietro Gavazzi)\n",
    "liste = merged.index.copy()\n",
    "liste = list(liste)\n",
    "\n",
    "for i in np.unique(merged.index):\n",
    "    liste.remove(i)\n",
    "\n",
    "## Change name to make it look nicer in the code!\n",
    "processed = merged.drop(np.unique(liste))\n",
    "\n",
    "print(\"Shape of the merged data set: \", processed.shape)\n",
    "print(\"with %d patients and %d features\" % (processed.shape[0], processed.shape[1]))\n",
    "\n",
    "processed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['NP_958782', 'NP_958785', 'NP_958786', 'NP_000436', 'NP_958781',\n",
       "       'NP_958780', 'NP_958783', 'NP_958784', 'NP_112598', 'NP_001611',\n",
       "       ...\n",
       "       'NP_001193600', 'NP_061134', 'NP_932347', 'NP_003593', 'NP_997203',\n",
       "       'NP_001191293', 'NP_775791', 'NP_004065', 'NP_068752', 'NP_219494'],\n",
       "      dtype='object', length=12553)"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Numerical data for the algorithm, NP_xx/XP_xx are protein identifiers from RefSeq database\n",
    "X = processed.loc[:,[x for x in processed.columns if bool(re.search(\"NP_|XP_|YP_\",x)) == True]]\n",
    "Y = pd.get_dummies(processed.drop(X.columns, axis=1)['Integrated Clusters (with PAM50)'], prefix=\"PAM50\")\n",
    "\n",
    "## Select only the PAM50 proteins - known panel of genes used for breast cancer subtype prediction\n",
    "# processed_numerical_p50 = processed_numerical.iloc[:,processed_numerical.columns.isin(pam50['RefSeqProteinID'])]\n",
    "# processed_numerical_p50.head()\n",
    "\n",
    "X.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PAM50_1</th>\n",
       "      <th>PAM50_2</th>\n",
       "      <th>PAM50_3</th>\n",
       "      <th>PAM50_4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0CM</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0D2</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0EQ</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0EV</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0EX</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              PAM50_1  PAM50_2  PAM50_3  PAM50_4\n",
       "TCGA-A2-A0CM        0        1        0        0\n",
       "TCGA-A2-A0D2        0        1        0        0\n",
       "TCGA-A2-A0EQ        0        0        0        1\n",
       "TCGA-A2-A0EV        0        0        1        0\n",
       "TCGA-A2-A0EX        0        0        1        0"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Y.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn [64], line 2\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[39m# Save the data\u001b[39;00m\n\u001b[1;32m----> 2\u001b[0m torch\u001b[39m.\u001b[39msave(X, \u001b[39m'\u001b[39m\u001b[39m./Results/X\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m      3\u001b[0m torch\u001b[39m.\u001b[39msave(Y, \u001b[39m'\u001b[39m\u001b[39m./Results/Y\u001b[39m\u001b[39m'\u001b[39m)\n",
      "File \u001b[1;32mc:\\Python310\\lib\\site-packages\\torch\\serialization.py:423\u001b[0m, in \u001b[0;36msave\u001b[1;34m(obj, f, pickle_module, pickle_protocol, _use_new_zipfile_serialization)\u001b[0m\n\u001b[0;32m    421\u001b[0m \u001b[39mif\u001b[39;00m _use_new_zipfile_serialization:\n\u001b[0;32m    422\u001b[0m     \u001b[39mwith\u001b[39;00m _open_zipfile_writer(f) \u001b[39mas\u001b[39;00m opened_zipfile:\n\u001b[1;32m--> 423\u001b[0m         _save(obj, opened_zipfile, pickle_module, pickle_protocol)\n\u001b[0;32m    424\u001b[0m         \u001b[39mreturn\u001b[39;00m\n\u001b[0;32m    425\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[1;32mc:\\Python310\\lib\\site-packages\\torch\\serialization.py:635\u001b[0m, in \u001b[0;36m_save\u001b[1;34m(obj, zip_file, pickle_module, pickle_protocol)\u001b[0m\n\u001b[0;32m    633\u001b[0m pickler \u001b[39m=\u001b[39m pickle_module\u001b[39m.\u001b[39mPickler(data_buf, protocol\u001b[39m=\u001b[39mpickle_protocol)\n\u001b[0;32m    634\u001b[0m pickler\u001b[39m.\u001b[39mpersistent_id \u001b[39m=\u001b[39m persistent_id\n\u001b[1;32m--> 635\u001b[0m pickler\u001b[39m.\u001b[39;49mdump(obj)\n\u001b[0;32m    636\u001b[0m data_value \u001b[39m=\u001b[39m data_buf\u001b[39m.\u001b[39mgetvalue()\n\u001b[0;32m    637\u001b[0m zip_file\u001b[39m.\u001b[39mwrite_record(\u001b[39m'\u001b[39m\u001b[39mdata.pkl\u001b[39m\u001b[39m'\u001b[39m, data_value, \u001b[39mlen\u001b[39m(data_value))\n",
      "File \u001b[1;32mc:\\Python310\\lib\\site-packages\\torch\\serialization.py:590\u001b[0m, in \u001b[0;36m_save.<locals>.persistent_id\u001b[1;34m(obj)\u001b[0m\n\u001b[0;32m    584\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mpersistent_id\u001b[39m(obj):\n\u001b[0;32m    585\u001b[0m     \u001b[39m# FIXME: the docs say that persistent_id should only return a string\u001b[39;00m\n\u001b[0;32m    586\u001b[0m     \u001b[39m# but torch store returns tuples. This works only in the binary protocol\u001b[39;00m\n\u001b[0;32m    587\u001b[0m     \u001b[39m# see\u001b[39;00m\n\u001b[0;32m    588\u001b[0m     \u001b[39m# https://docs.python.org/2/library/pickle.html#pickling-and-unpickling-external-objects\u001b[39;00m\n\u001b[0;32m    589\u001b[0m     \u001b[39m# https://github.com/python/cpython/blob/master/Lib/pickle.py#L527-L537\u001b[39;00m\n\u001b[1;32m--> 590\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39;49m(obj, torch\u001b[39m.\u001b[39;49mstorage\u001b[39m.\u001b[39;49mTypedStorage) \u001b[39mor\u001b[39;00m torch\u001b[39m.\u001b[39mis_storage(obj):\n\u001b[0;32m    592\u001b[0m         \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(obj, torch\u001b[39m.\u001b[39mstorage\u001b[39m.\u001b[39mTypedStorage):\n\u001b[0;32m    593\u001b[0m             \u001b[39m# TODO: Once we decide to break serialization FC, this case\u001b[39;00m\n\u001b[0;32m    594\u001b[0m             \u001b[39m# can be deleted\u001b[39;00m\n\u001b[0;32m    595\u001b[0m             storage \u001b[39m=\u001b[39m obj\u001b[39m.\u001b[39m_storage\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Save the data\n",
    "torch.save(X, './Results/X')\n",
    "torch.save(Y, './Results/Y')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## 3. Data Engineering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Read the processed data set\n",
    "X = torch.load('./Results/X')\n",
    "Y = torch.load('./Results/Y')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NP_958782</th>\n",
       "      <th>NP_958785</th>\n",
       "      <th>NP_958786</th>\n",
       "      <th>NP_000436</th>\n",
       "      <th>NP_958781</th>\n",
       "      <th>NP_958780</th>\n",
       "      <th>NP_958783</th>\n",
       "      <th>NP_958784</th>\n",
       "      <th>NP_112598</th>\n",
       "      <th>NP_001611</th>\n",
       "      <th>...</th>\n",
       "      <th>NP_001193600</th>\n",
       "      <th>NP_061134</th>\n",
       "      <th>NP_932347</th>\n",
       "      <th>NP_003593</th>\n",
       "      <th>NP_997203</th>\n",
       "      <th>NP_001191293</th>\n",
       "      <th>NP_775791</th>\n",
       "      <th>NP_004065</th>\n",
       "      <th>NP_068752</th>\n",
       "      <th>NP_219494</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0CM</th>\n",
       "      <td>0.683404</td>\n",
       "      <td>0.694424</td>\n",
       "      <td>0.698098</td>\n",
       "      <td>0.687077</td>\n",
       "      <td>0.687077</td>\n",
       "      <td>0.698098</td>\n",
       "      <td>0.698098</td>\n",
       "      <td>0.698098</td>\n",
       "      <td>-2.652150</td>\n",
       "      <td>-0.984373</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.153614</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0D2</th>\n",
       "      <td>0.107491</td>\n",
       "      <td>0.104164</td>\n",
       "      <td>0.107491</td>\n",
       "      <td>0.097512</td>\n",
       "      <td>0.104164</td>\n",
       "      <td>0.104164</td>\n",
       "      <td>0.104164</td>\n",
       "      <td>0.104164</td>\n",
       "      <td>-0.880454</td>\n",
       "      <td>-1.512473</td>\n",
       "      <td>...</td>\n",
       "      <td>0.919136</td>\n",
       "      <td>-1.648856</td>\n",
       "      <td>0.832649</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-8.324969</td>\n",
       "      <td>-4.679219</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.106650</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-6.941181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0EQ</th>\n",
       "      <td>-0.912670</td>\n",
       "      <td>-0.927979</td>\n",
       "      <td>-0.927979</td>\n",
       "      <td>-0.931806</td>\n",
       "      <td>-0.927979</td>\n",
       "      <td>-0.927979</td>\n",
       "      <td>-0.927979</td>\n",
       "      <td>-0.927979</td>\n",
       "      <td>-3.071151</td>\n",
       "      <td>-2.278943</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.801685</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.802310</td>\n",
       "      <td>-6.373934</td>\n",
       "      <td>-1.123160</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0EV</th>\n",
       "      <td>0.452986</td>\n",
       "      <td>0.472590</td>\n",
       "      <td>0.472590</td>\n",
       "      <td>0.458587</td>\n",
       "      <td>0.472590</td>\n",
       "      <td>0.472590</td>\n",
       "      <td>0.472590</td>\n",
       "      <td>0.472590</td>\n",
       "      <td>-0.742871</td>\n",
       "      <td>1.811277</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.966177</td>\n",
       "      <td>-1.471027</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.474013</td>\n",
       "      <td>-12.278546</td>\n",
       "      <td>-10.337729</td>\n",
       "      <td>-0.653251</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0EX</th>\n",
       "      <td>1.185108</td>\n",
       "      <td>1.192612</td>\n",
       "      <td>1.188860</td>\n",
       "      <td>1.185108</td>\n",
       "      <td>1.200116</td>\n",
       "      <td>1.188860</td>\n",
       "      <td>1.188860</td>\n",
       "      <td>1.192612</td>\n",
       "      <td>1.046289</td>\n",
       "      <td>2.138081</td>\n",
       "      <td>...</td>\n",
       "      <td>1.451490</td>\n",
       "      <td>-2.018981</td>\n",
       "      <td>0.877456</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-6.101005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-1.726336</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 12553 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              NP_958782  NP_958785  NP_958786  NP_000436  NP_958781  \\\n",
       "TCGA-A2-A0CM   0.683404   0.694424   0.698098   0.687077   0.687077   \n",
       "TCGA-A2-A0D2   0.107491   0.104164   0.107491   0.097512   0.104164   \n",
       "TCGA-A2-A0EQ  -0.912670  -0.927979  -0.927979  -0.931806  -0.927979   \n",
       "TCGA-A2-A0EV   0.452986   0.472590   0.472590   0.458587   0.472590   \n",
       "TCGA-A2-A0EX   1.185108   1.192612   1.188860   1.185108   1.200116   \n",
       "\n",
       "              NP_958780  NP_958783  NP_958784  NP_112598  NP_001611  ...  \\\n",
       "TCGA-A2-A0CM   0.698098   0.698098   0.698098  -2.652150  -0.984373  ...   \n",
       "TCGA-A2-A0D2   0.104164   0.104164   0.104164  -0.880454  -1.512473  ...   \n",
       "TCGA-A2-A0EQ  -0.927979  -0.927979  -0.927979  -3.071151  -2.278943  ...   \n",
       "TCGA-A2-A0EV   0.472590   0.472590   0.472590  -0.742871   1.811277  ...   \n",
       "TCGA-A2-A0EX   1.188860   1.188860   1.192612   1.046289   2.138081  ...   \n",
       "\n",
       "              NP_001193600  NP_061134  NP_932347  NP_003593  NP_997203  \\\n",
       "TCGA-A2-A0CM           NaN        NaN   1.153614        NaN        NaN   \n",
       "TCGA-A2-A0D2      0.919136  -1.648856   0.832649        NaN  -8.324969   \n",
       "TCGA-A2-A0EQ     -0.801685        NaN        NaN   3.802310  -6.373934   \n",
       "TCGA-A2-A0EV     -4.966177  -1.471027        NaN  -0.474013 -12.278546   \n",
       "TCGA-A2-A0EX      1.451490  -2.018981   0.877456        NaN        NaN   \n",
       "\n",
       "              NP_001191293  NP_775791  NP_004065  NP_068752  NP_219494  \n",
       "TCGA-A2-A0CM           NaN        NaN        NaN        NaN        NaN  \n",
       "TCGA-A2-A0D2     -4.679219        NaN  -1.106650        NaN  -6.941181  \n",
       "TCGA-A2-A0EQ     -1.123160        NaN        NaN        NaN        NaN  \n",
       "TCGA-A2-A0EV    -10.337729  -0.653251        NaN        NaN        NaN  \n",
       "TCGA-A2-A0EX     -6.101005        NaN  -1.726336        NaN        NaN  \n",
       "\n",
       "[5 rows x 12553 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PAM50_1</th>\n",
       "      <th>PAM50_2</th>\n",
       "      <th>PAM50_3</th>\n",
       "      <th>PAM50_4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0CM</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0D2</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0EQ</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0EV</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-A2-A0EX</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              PAM50_1  PAM50_2  PAM50_3  PAM50_4\n",
       "TCGA-A2-A0CM        0        1        0        0\n",
       "TCGA-A2-A0D2        0        1        0        0\n",
       "TCGA-A2-A0EQ        0        0        0        1\n",
       "TCGA-A2-A0EV        0        0        1        0\n",
       "TCGA-A2-A0EX        0        0        1        0"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Y.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### 3.1 Missing value process\n",
    "We drop all the columns with nan values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "nan_counts = np.sum(X.isna(), axis=0)\n",
    "col_to_drop = nan_counts[nan_counts != 0].index\n",
    "X.drop(col_to_drop, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(74, 8023)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### 3.2 Data Scaling\n",
    "We use the standardization for each feature of the data set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(74, 8023)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "col_names = X.columns\n",
    "idx = X.index\n",
    "scaler = StandardScaler()\n",
    "X_scaled = pd.DataFrame(scaler.fit_transform(X), columns=col_names, index=idx)\n",
    "X_scaled.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## 4. Basic model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### 4.1 K-Means model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Python310\\lib\\site-packages\\sklearn\\cluster\\_kmeans.py:1318: FutureWarning: algorithm='full' is deprecated, it will be removed in 1.3. Using 'lloyd' instead.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "kmeans = KMeans(n_clusters=4, algorithm='full').fit(X_scaled)\n",
    "pred0 = kmeans.labels_"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### 4.2 Evaluation Metric"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "# https://medium.com/data-science-in-your-pocket/calculating-precision-recall-for-multi-class-classification-9055931ee229 \n",
    "\n",
    "\n",
    "permutation_matrices = []\n",
    "for i in itertools.permutations([0, 1, 2, 3]):\n",
    "    matrix = np.zeros((4, 4))\n",
    "    for j in range(len(matrix)):\n",
    "        matrix[j][i[j]] = 1\n",
    "\n",
    "    permutation_matrices.append(matrix)\n",
    "\n",
    "\n",
    "\n",
    "def get_dataframe(y_array_pred, y_true):\n",
    "    pred = pd.DataFrame(pd.get_dummies(y_array_pred))\n",
    "    pred.columns = ['cluster_' + str(x) for x in range(1, 5)]\n",
    "    pred = pred.set_index(y_true.index)\n",
    "    return(pred)\n",
    "\n",
    "\n",
    "\n",
    "def calculate_mean_prec(matrix):\n",
    "    matrix = np.array(matrix)\n",
    "    assert(matrix.shape[0]==matrix.shape[1]), f\"matrix must be squared, same number of true labels {matrix.shape[0]} as predicted{matrix.shape[1]}\"\n",
    "\n",
    "    prec = np.zeros(len(matrix))\n",
    "\n",
    "    for i in range(len(matrix)):\n",
    "        true_pos_classified_pos = matrix[i][i]\n",
    "        classified_pos= np.sum(matrix[i])\n",
    "        prec[i] = true_pos_classified_pos/classified_pos\n",
    "    \n",
    "    return np.mean(prec)\n",
    "\n",
    "def calculate_mean_recall(matrix):\n",
    "    matrix = np.array(matrix)\n",
    "    assert(matrix.shape[0]==matrix.shape[1]), f\"matrix must be squared, same number of true labels {matrix.shape[0]} as predicted{matrix.shape[1]}\"\n",
    "\n",
    "    rec = np.zeros(len(matrix.T))\n",
    "\n",
    "    for i in range(len(matrix.T)):\n",
    "        true_pos_classified_pos = matrix.T[i][i]\n",
    "        true_pos= np.sum(matrix.T[i])\n",
    "        rec[i] = true_pos_classified_pos/true_pos\n",
    "\n",
    "    return np.mean(rec)\n",
    "\n",
    "\n",
    "def model_eval(y_true, y_pred):\n",
    "\n",
    "    length = len(y_true.T)\n",
    "    assert (y_pred.shape==y_true.shape), f\"expected same shape of colums for both y_true: {y_true.shape} and y_pred {y_pred.shape}\"\n",
    "    \n",
    "    matrix = np.zeros((length, length))\n",
    "\n",
    "    indi = 0\n",
    "    for i in y_pred.T.index:\n",
    "        indj = 0\n",
    "        for j in y_true.T.index:\n",
    "            matrix[indi][indj] += np.array(y_pred[i]) @ np.array(y_true[j])\n",
    "            indj+=1\n",
    "        indi+=1\n",
    "\n",
    "    \n",
    "    matrix = pd.DataFrame(matrix, columns=y_true.T.index, index=y_pred.T.index)\n",
    "\n",
    "\n",
    "    best_perm = None\n",
    "    best_Fscore = 0\n",
    "\n",
    "    ## version 3.0\n",
    "    for permu_mat in permutation_matrices:\n",
    "        M = permu_mat@matrix\n",
    "\n",
    "        prec = calculate_mean_prec(M)\n",
    "        rec = calculate_mean_recall(M)\n",
    "\n",
    "        Fscore = 2/(1/prec + 1/rec)\n",
    "        \n",
    "        if Fscore>best_Fscore:\n",
    "            best_perm = permu_mat\n",
    "            best_Fscore = Fscore\n",
    "    \n",
    "    new_matrix = best_perm@matrix\n",
    "\n",
    "    indexes = []\n",
    "    for i in range(len(best_perm)):\n",
    "        for j in range(len(best_perm)):\n",
    "            if best_perm[i][j]==1:\n",
    "                indexes.append(matrix.index[j])\n",
    "\n",
    "\n",
    "    new_matrix.index = indexes\n",
    "\n",
    "    return new_matrix, best_Fscore\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PAM50_1</th>\n",
       "      <th>PAM50_2</th>\n",
       "      <th>PAM50_3</th>\n",
       "      <th>PAM50_4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cluster_1</th>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_3</th>\n",
       "      <td>4.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_2</th>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           PAM50_1  PAM50_2  PAM50_3  PAM50_4\n",
       "cluster_1      4.0      0.0      1.0      3.0\n",
       "cluster_3      4.0      9.0      3.0      6.0\n",
       "cluster_4      0.0      4.0     13.0      3.0\n",
       "cluster_2      2.0      5.0      6.0     11.0"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_pred = get_dataframe(pred0, Y)\n",
    "new_matrix, best_Fscore = model_eval(Y, y_pred)\n",
    "new_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.4949402509708698"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "best_Fscore"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## 5. Model improvement"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### 5.1 Feature selection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "501\n",
      "1\n",
      "501\n",
      "2\n",
      "501\n",
      "3\n",
      "501\n",
      "4\n",
      "501\n",
      "5\n",
      "501\n",
      "6\n",
      "501\n",
      "7\n",
      "501\n",
      "8\n",
      "501\n",
      "9\n",
      "501\n",
      "10\n",
      "501\n",
      "11\n",
      "501\n",
      "12\n",
      "501\n",
      "13\n",
      "501\n",
      "14\n",
      "501\n",
      "15\n",
      "501\n",
      "16\n",
      "501\n",
      "17\n",
      "501\n",
      "18\n",
      "501\n",
      "19\n",
      "501\n",
      "20\n",
      "501\n",
      "21\n",
      "501\n",
      "22\n",
      "501\n",
      "23\n",
      "501\n",
      "24\n",
      "501\n",
      "25\n",
      "501\n",
      "26\n",
      "501\n",
      "27\n",
      "501\n",
      "28\n",
      "501\n",
      "29\n",
      "501\n",
      "30\n",
      "501\n",
      "31\n",
      "501\n",
      "32\n",
      "501\n",
      "33\n",
      "501\n",
      "34\n",
      "501\n",
      "35\n",
      "501\n",
      "36\n",
      "501\n",
      "37\n",
      "501\n",
      "38\n",
      "501\n",
      "39\n",
      "501\n",
      "40\n",
      "501\n",
      "41\n",
      "501\n",
      "42\n",
      "501\n",
      "43\n",
      "501\n",
      "44\n",
      "501\n",
      "45\n",
      "501\n",
      "46\n",
      "501\n",
      "47\n",
      "501\n",
      "48\n",
      "501\n",
      "49\n",
      "501\n",
      "50\n",
      "501\n",
      "51\n",
      "501\n",
      "52\n",
      "501\n",
      "53\n",
      "501\n",
      "54\n",
      "501\n",
      "55\n",
      "501\n",
      "56\n",
      "501\n",
      "57\n",
      "501\n",
      "58\n",
      "501\n",
      "59\n",
      "501\n",
      "60\n",
      "501\n",
      "61\n",
      "501\n",
      "62\n",
      "501\n",
      "63\n",
      "501\n",
      "64\n",
      "501\n",
      "65\n",
      "501\n",
      "66\n",
      "501\n",
      "67\n",
      "501\n",
      "68\n",
      "501\n",
      "69\n",
      "501\n",
      "70\n",
      "501\n",
      "71\n",
      "501\n",
      "72\n",
      "501\n",
      "73\n",
      "501\n",
      "74\n",
      "501\n",
      "75\n",
      "501\n",
      "76\n",
      "501\n",
      "77\n",
      "501\n",
      "78\n",
      "501\n",
      "79\n",
      "501\n",
      "80\n",
      "501\n",
      "81\n",
      "501\n",
      "82\n",
      "501\n",
      "83\n",
      "501\n",
      "84\n",
      "501\n",
      "85\n",
      "501\n",
      "86\n",
      "501\n",
      "87\n",
      "501\n",
      "88\n",
      "501\n",
      "89\n",
      "501\n",
      "90\n",
      "501\n",
      "91\n",
      "501\n",
      "92\n",
      "501\n",
      "93\n",
      "501\n",
      "94\n",
      "501\n",
      "95\n",
      "501\n",
      "96\n",
      "501\n",
      "97\n",
      "501\n",
      "98\n",
      "501\n",
      "99\n",
      "501\n",
      "100\n",
      "501\n",
      "101\n",
      "501\n",
      "102\n",
      "501\n",
      "103\n",
      "501\n",
      "104\n",
      "501\n",
      "105\n",
      "501\n",
      "106\n",
      "501\n",
      "107\n",
      "501\n",
      "108\n",
      "501\n",
      "109\n",
      "501\n",
      "110\n",
      "501\n",
      "111\n",
      "501\n",
      "112\n",
      "501\n",
      "113\n",
      "501\n",
      "114\n",
      "501\n",
      "115\n",
      "501\n",
      "116\n",
      "501\n",
      "117\n",
      "501\n",
      "118\n",
      "501\n",
      "119\n",
      "501\n",
      "120\n",
      "501\n",
      "121\n",
      "501\n",
      "122\n",
      "501\n",
      "123\n",
      "501\n",
      "124\n",
      "501\n",
      "125\n",
      "501\n",
      "126\n",
      "501\n",
      "127\n",
      "501\n",
      "128\n",
      "501\n",
      "129\n",
      "501\n",
      "130\n",
      "501\n",
      "131\n",
      "501\n",
      "132\n",
      "501\n",
      "133\n",
      "501\n",
      "134\n",
      "501\n",
      "135\n",
      "501\n",
      "136\n",
      "501\n",
      "137\n",
      "501\n",
      "138\n",
      "501\n",
      "139\n",
      "501\n",
      "140\n",
      "501\n",
      "141\n",
      "501\n",
      "142\n",
      "501\n",
      "143\n",
      "501\n",
      "144\n",
      "501\n",
      "145\n",
      "501\n",
      "146\n",
      "501\n",
      "147\n",
      "501\n",
      "148\n",
      "501\n",
      "149\n",
      "501\n",
      "150\n",
      "501\n",
      "151\n",
      "501\n",
      "152\n",
      "501\n",
      "153\n",
      "501\n",
      "154\n",
      "501\n",
      "155\n",
      "501\n",
      "156\n",
      "501\n",
      "157\n",
      "501\n",
      "158\n",
      "501\n",
      "159\n",
      "501\n",
      "160\n",
      "501\n",
      "161\n",
      "501\n",
      "162\n",
      "501\n",
      "163\n",
      "501\n",
      "164\n",
      "501\n",
      "165\n",
      "501\n",
      "166\n",
      "501\n",
      "167\n",
      "501\n",
      "168\n",
      "501\n",
      "169\n",
      "501\n",
      "170\n",
      "501\n",
      "171\n",
      "501\n",
      "172\n",
      "501\n",
      "173\n",
      "501\n",
      "174\n",
      "501\n",
      "175\n",
      "501\n",
      "176\n",
      "501\n",
      "177\n",
      "501\n",
      "178\n",
      "501\n",
      "179\n",
      "501\n",
      "180\n",
      "501\n",
      "181\n",
      "501\n",
      "182\n",
      "501\n",
      "183\n",
      "501\n",
      "184\n",
      "501\n",
      "185\n",
      "501\n",
      "186\n",
      "501\n",
      "187\n",
      "501\n",
      "188\n",
      "501\n",
      "189\n",
      "501\n",
      "190\n",
      "501\n",
      "191\n",
      "501\n",
      "192\n",
      "501\n",
      "193\n",
      "501\n",
      "194\n",
      "501\n",
      "195\n",
      "501\n",
      "196\n",
      "501\n",
      "197\n",
      "501\n",
      "198\n",
      "501\n",
      "199\n",
      "501\n",
      "200\n",
      "501\n",
      "201\n",
      "501\n",
      "202\n",
      "501\n",
      "203\n",
      "501\n",
      "204\n",
      "501\n",
      "205\n",
      "501\n",
      "206\n",
      "501\n",
      "207\n",
      "501\n",
      "208\n",
      "501\n",
      "209\n",
      "501\n",
      "210\n",
      "501\n",
      "211\n",
      "501\n",
      "212\n",
      "501\n",
      "213\n",
      "501\n",
      "214\n",
      "501\n",
      "215\n",
      "501\n",
      "216\n",
      "501\n",
      "217\n",
      "501\n",
      "218\n",
      "501\n",
      "219\n",
      "501\n",
      "220\n",
      "501\n",
      "221\n",
      "501\n",
      "222\n",
      "501\n",
      "223\n",
      "501\n",
      "224\n",
      "501\n",
      "225\n",
      "501\n",
      "226\n",
      "501\n",
      "227\n",
      "501\n",
      "228\n",
      "501\n",
      "229\n",
      "501\n",
      "230\n",
      "501\n",
      "231\n",
      "501\n",
      "232\n",
      "501\n",
      "233\n",
      "501\n",
      "234\n",
      "501\n",
      "235\n",
      "501\n",
      "236\n",
      "501\n",
      "237\n",
      "501\n",
      "238\n",
      "501\n",
      "239\n",
      "501\n",
      "240\n",
      "501\n",
      "241\n",
      "501\n",
      "242\n",
      "501\n",
      "243\n",
      "501\n",
      "244\n",
      "501\n",
      "245\n",
      "501\n",
      "246\n",
      "501\n",
      "247\n",
      "501\n",
      "248\n",
      "501\n",
      "249\n",
      "501\n",
      "250\n",
      "501\n",
      "251\n",
      "501\n",
      "252\n",
      "501\n",
      "253\n",
      "501\n",
      "254\n",
      "501\n",
      "255\n",
      "501\n",
      "256\n",
      "501\n",
      "257\n",
      "501\n",
      "258\n",
      "501\n",
      "259\n",
      "501\n",
      "260\n",
      "501\n",
      "261\n",
      "501\n",
      "262\n",
      "501\n",
      "263\n",
      "501\n",
      "264\n",
      "501\n",
      "265\n",
      "501\n",
      "266\n",
      "501\n",
      "267\n",
      "501\n",
      "268\n",
      "501\n",
      "269\n",
      "501\n",
      "270\n",
      "501\n",
      "271\n",
      "501\n",
      "272\n",
      "501\n",
      "273\n",
      "501\n",
      "274\n",
      "501\n",
      "275\n",
      "501\n",
      "276\n",
      "501\n",
      "277\n",
      "501\n",
      "278\n",
      "501\n",
      "279\n",
      "501\n",
      "280\n",
      "501\n",
      "281\n",
      "501\n",
      "282\n",
      "501\n",
      "283\n",
      "501\n",
      "284\n",
      "501\n",
      "285\n",
      "501\n",
      "286\n",
      "501\n",
      "287\n",
      "501\n",
      "288\n",
      "501\n",
      "289\n",
      "501\n",
      "290\n",
      "501\n",
      "291\n",
      "501\n",
      "292\n",
      "501\n",
      "293\n",
      "501\n",
      "294\n",
      "501\n",
      "295\n",
      "501\n",
      "296\n",
      "501\n",
      "297\n",
      "501\n",
      "298\n",
      "501\n",
      "299\n",
      "501\n",
      "300\n",
      "501\n",
      "301\n",
      "501\n",
      "302\n",
      "501\n",
      "303\n",
      "501\n",
      "304\n",
      "501\n",
      "305\n",
      "501\n",
      "306\n",
      "501\n",
      "307\n",
      "501\n",
      "308\n",
      "501\n",
      "309\n",
      "501\n",
      "310\n",
      "501\n",
      "311\n",
      "501\n",
      "312\n",
      "501\n",
      "313\n",
      "501\n",
      "314\n",
      "501\n",
      "315\n",
      "501\n",
      "316\n",
      "501\n",
      "317\n",
      "501\n",
      "318\n",
      "501\n",
      "319\n",
      "501\n",
      "320\n",
      "501\n",
      "321\n",
      "501\n",
      "322\n",
      "501\n",
      "323\n",
      "501\n",
      "324\n",
      "501\n",
      "325\n",
      "501\n",
      "326\n",
      "501\n",
      "327\n",
      "501\n",
      "328\n",
      "501\n",
      "329\n",
      "501\n",
      "330\n",
      "501\n",
      "331\n",
      "501\n",
      "332\n",
      "501\n",
      "333\n",
      "501\n",
      "334\n",
      "501\n",
      "335\n",
      "501\n",
      "336\n",
      "501\n",
      "337\n",
      "501\n",
      "338\n",
      "501\n",
      "339\n",
      "501\n",
      "340\n",
      "501\n",
      "341\n",
      "501\n",
      "342\n",
      "501\n",
      "343\n",
      "501\n",
      "344\n",
      "501\n",
      "345\n",
      "501\n",
      "346\n",
      "501\n",
      "347\n",
      "501\n",
      "348\n",
      "501\n",
      "349\n",
      "501\n",
      "350\n",
      "501\n",
      "351\n",
      "501\n",
      "352\n",
      "501\n",
      "353\n",
      "501\n",
      "354\n",
      "501\n",
      "355\n",
      "501\n",
      "356\n",
      "501\n",
      "357\n",
      "501\n",
      "358\n",
      "501\n",
      "359\n",
      "501\n",
      "360\n",
      "501\n",
      "361\n",
      "501\n",
      "362\n",
      "501\n",
      "363\n",
      "501\n",
      "364\n",
      "501\n",
      "365\n",
      "501\n",
      "366\n",
      "501\n",
      "367\n",
      "501\n",
      "368\n",
      "501\n",
      "369\n",
      "501\n",
      "370\n",
      "501\n",
      "371\n",
      "501\n",
      "372\n",
      "501\n",
      "373\n",
      "501\n",
      "374\n",
      "501\n",
      "375\n",
      "501\n",
      "376\n",
      "501\n",
      "377\n",
      "501\n",
      "378\n",
      "501\n",
      "379\n",
      "501\n",
      "380\n",
      "501\n",
      "381\n",
      "501\n",
      "382\n",
      "501\n",
      "383\n",
      "501\n",
      "384\n",
      "501\n",
      "385\n",
      "501\n",
      "386\n",
      "501\n",
      "387\n",
      "501\n",
      "388\n",
      "501\n",
      "389\n",
      "501\n",
      "390\n",
      "501\n",
      "391\n",
      "501\n",
      "392\n",
      "501\n",
      "393\n",
      "501\n",
      "394\n",
      "501\n",
      "395\n",
      "501\n",
      "396\n",
      "501\n",
      "397\n",
      "501\n",
      "398\n",
      "501\n",
      "399\n",
      "501\n",
      "400\n",
      "501\n",
      "401\n",
      "501\n",
      "402\n",
      "501\n",
      "403\n",
      "501\n",
      "404\n",
      "501\n",
      "405\n",
      "501\n",
      "406\n",
      "501\n",
      "407\n",
      "501\n",
      "408\n",
      "501\n",
      "409\n",
      "501\n",
      "410\n",
      "501\n",
      "411\n",
      "501\n",
      "412\n",
      "501\n",
      "413\n",
      "501\n",
      "414\n",
      "501\n",
      "415\n",
      "501\n",
      "416\n",
      "501\n",
      "417\n",
      "501\n",
      "418\n",
      "501\n",
      "419\n",
      "501\n",
      "420\n",
      "501\n",
      "421\n",
      "501\n",
      "422\n",
      "501\n",
      "423\n",
      "501\n",
      "424\n",
      "501\n",
      "425\n",
      "501\n",
      "426\n",
      "501\n",
      "427\n",
      "501\n",
      "428\n",
      "501\n",
      "429\n",
      "501\n",
      "430\n",
      "501\n",
      "431\n",
      "501\n",
      "432\n",
      "501\n",
      "433\n",
      "501\n",
      "434\n",
      "501\n",
      "435\n",
      "501\n",
      "436\n",
      "501\n",
      "437\n",
      "501\n",
      "438\n",
      "501\n",
      "439\n",
      "501\n",
      "440\n",
      "501\n",
      "441\n",
      "501\n",
      "442\n",
      "501\n",
      "443\n",
      "501\n",
      "444\n",
      "501\n",
      "445\n",
      "501\n",
      "446\n",
      "501\n",
      "447\n",
      "501\n",
      "448\n",
      "501\n",
      "449\n",
      "501\n",
      "450\n",
      "501\n",
      "451\n",
      "501\n",
      "452\n",
      "501\n",
      "453\n",
      "501\n",
      "454\n",
      "501\n",
      "455\n",
      "501\n",
      "456\n",
      "501\n",
      "457\n",
      "501\n",
      "458\n",
      "501\n",
      "459\n",
      "501\n",
      "460\n",
      "501\n",
      "461\n",
      "501\n",
      "462\n",
      "501\n",
      "463\n",
      "501\n",
      "464\n",
      "501\n",
      "465\n",
      "501\n",
      "466\n",
      "501\n",
      "467\n",
      "501\n",
      "468\n",
      "501\n",
      "469\n",
      "501\n",
      "470\n",
      "501\n",
      "471\n",
      "501\n",
      "472\n",
      "501\n",
      "473\n",
      "501\n",
      "474\n",
      "501\n",
      "475\n",
      "501\n",
      "476\n",
      "501\n",
      "477\n",
      "501\n",
      "478\n",
      "501\n",
      "479\n",
      "501\n",
      "480\n",
      "501\n",
      "481\n",
      "501\n",
      "482\n",
      "501\n",
      "483\n",
      "501\n",
      "484\n",
      "501\n",
      "485\n",
      "501\n",
      "486\n",
      "501\n",
      "487\n",
      "501\n",
      "488\n",
      "501\n",
      "489\n",
      "501\n",
      "490\n",
      "501\n",
      "491\n",
      "501\n",
      "492\n",
      "501\n",
      "493\n",
      "501\n",
      "494\n",
      "501\n",
      "495\n",
      "501\n",
      "496\n",
      "501\n",
      "497\n",
      "501\n",
      "498\n",
      "501\n",
      "499\n",
      "501\n",
      "500\n",
      "501\n",
      "0.5023116440802425\n",
      "NP_004332\n"
     ]
    }
   ],
   "source": [
    "Fscor_list = []\n",
    "max_Fscore = 0\n",
    "feature_selected = None\n",
    "\n",
    "for i in range(X_scaled.shape[1]):\n",
    "    print(i)\n",
    "    print(X_scaled.shape[1])\n",
    "    X_reduced = np.array(X_scaled.iloc[:, i]).reshape(-1, 1)\n",
    "    k_means = KMeans(n_clusters=4).fit(X_reduced)\n",
    "    pred_ = k_means.labels_\n",
    "    y_pred_ = get_dataframe(pred_, Y)\n",
    "    _, Fscore = model_eval(Y, y_pred_)\n",
    "    Fscor_list.append(Fscore)\n",
    "\n",
    "    if Fscore > max_Fscore:\n",
    "        max_Fscore = Fscore\n",
    "        feature_selected = X_scaled.columns[i]\n",
    "\n",
    "print(max_Fscore)\n",
    "print(feature_selected)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PAM50_1</th>\n",
       "      <th>PAM50_2</th>\n",
       "      <th>PAM50_3</th>\n",
       "      <th>PAM50_4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cluster_1</th>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cluster_4</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           PAM50_1  PAM50_2  PAM50_3  PAM50_4\n",
       "cluster_1      8.0      3.0     11.0      8.0\n",
       "cluster_3      1.0      7.0      1.0      1.0\n",
       "cluster_2      0.0      1.0      4.0      1.0\n",
       "cluster_4      1.0      7.0      7.0     13.0"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_reduced = np.array(X_scaled[feature_selected]).reshape(-1, 1)\n",
    "k_means = KMeans(n_clusters=4).fit(X_reduced)\n",
    "pred = k_means.labels_\n",
    "y_pred = get_dataframe(pred, Y)\n",
    "error = model_eval(Y, y_pred)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "### 5.2 Random Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  },
  "vscode": {
   "interpreter": {
    "hash": "369f2c481f4da34e4445cda3fffd2e751bd1c4d706f27375911949ba6bb62e1c"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
